docs/tutorial/saxpy.rst

```diff
@@ -0,0 +1,780 @@
+  {
+      // ...
+  }
+
+  int main()
+  {
+      // ...
+
+      // Launch the kernel on the default stream.
+      saxpy_kernel<<<dim3(grid_size), dim3(block_size), 0, hipStreamDefault>>>(a, d_x, d_y, size);
+  }
+
+Analyze at the signature of the offloaded function:
+
+- ``__global__`` instructs the compiler to generate code for this function as an
+  entrypoint to a device program, such that it can be launched from the host.
+- The function does not return anything, because there is no trivial way to
+  construct a return channel of a parallel invocation. Device-side entrypoints
+  may not return a value, their results should be communicated using output
+  parameters.
+- Device-side functions are typically called compute kernels, or just kernels
+  for short. This is to distinguish them from non-graphics-related graphics
+  shaders, or just shaders for short.
+- Arguments are taken by value and all arguments shall be
+  `TriviallyCopyable <https://en.cppreference.com/w/cpp/named_req/TriviallyCopyable>`_,
+  meaning they should be `memcpy`-friendly. _(Imagine if they had custom copy
+  constructors. Where would that logic execute? On the host? On the device?)_
+  Pointer arguments are pointers to device memory, one typically backed by
+  VRAM.
+- We said that we'll be computing :math:`a\cdot x+y=z`, however we only pass
+  two pointers to the function. We'll be canonically reusing one of the inputs
+  as outputs.
+
+This function is launched from the host using a language extension often called
+the triple chevron syntax. Inside the angle brackets, provide the following.
+
+- The number of blocks to launch (our grid size)
+- The number of threads in a block (our block size)
+- The amount of shared memory to allocate by the host
+- The device stream to enqueue the operation on
+
+The block size and shared memory become important later in :doc:`reduction`. For
+now, a hardcoded ``256`` is a safe default for simple kernels such as this.
+Following the triple chevron is ordinary function argument passing.
+
+Look at how the kernel is implemented.
+
+.. code-block:: C++
+
+  __global__ void saxpy_kernel(const float a, const float* d_x, float* d_y, const unsigned int size)
+  {
+      // Compute the current thread's index in the grid.
+      const unsigned int global_idx = blockIdx.x * blockDim.x + threadIdx.x;
+
+      // The grid can be larger than the number of items in the vectors. Avoid out-of-bounds addressing.
+      if(global_idx < size)
+      {
+          d_y[global_idx] = a * d_x[global_idx] + d_y[global_idx];
+      }
+  }
+
+- The unique linear index identifying the thread is computed from the block id
+  the thread is a member of, the block's size and the id of the thread within
+  the block.
+- A check is made to avoid overindexing the input.
+- The useful part of the computation is carried out.
+
+Retrieval of the result from the device is done much like its dispatching:
```

Comment:
dispatching can mean different thing in context of HIP, can we change the wording here. Also the first memcpy should also be explained and here we can mention in simple terms that we are copying result back to host.

Reply:
Removed dispatching.